chained/water_impl.go

```diff
@@ -22,74 +23,105 @@ type waterImpl struct {
-	wasmAvailableAt := ptSetting(pc, "water_available_at")
-	transport := ptSetting(pc, "water_transport")
-	if wasm == nil && wasmAvailableAt != "" {
-		vc := newWaterVersionControl(dir)
-		cli := httpClient
-		if cli == nil {
-			cli = proxied.ChainedThenDirectThenFrontedClient(1*time.Minute, "")
-		}
-		downloader, err := newWaterWASMDownloader(strings.Split(wasmAvailableAt, ","), cli)
-		if err != nil {
-			return nil, log.Errorf("failed to create wasm downloader: %w", err)
-		}
 
-		r, err := vc.GetWASM(ctx, transport, downloader)
+		dialer, err := createDialer(ctx, wasm, transport)
 		if err != nil {
-			return nil, log.Errorf("failed to get wasm: %w", err)
-		}
-		defer r.Close()
-
-		b, err := io.ReadAll(r)
-		if err != nil {
-			return nil, log.Errorf("failed to read wasm: %w", err)
-		}
-
-		if len(b) == 0 {
-			return nil, log.Errorf("received empty wasm")
+			return nil, log.Errorf("failed to create dialer: %w", err)
 		}
+		d.dialer = dialer
+		return d, nil
+	}
 
-		wasm = b
+	if wasmAvailableAt != "" {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			log.Debugf("Loading WASM for %q. If not available, it should try to download from the following URLs: %+v. The file should be available here: %s", transport, strings.Split(wasmAvailableAt, ","), dir)
+			vc := newWaterVersionControl(dir)
+			cli := httpClient
+			if cli == nil {
+				cli = proxied.ChainedThenDirectThenFrontedClient(1*time.Minute, "")
+			}
+			downloader, err := newWaterWASMDownloader(strings.Split(wasmAvailableAt, ","), cli)
+			if err != nil {
+				d.downloadErr = log.Errorf("failed to create wasm downloader: %w", err)
+				return
+			}
+
+			r, err := vc.GetWASM(ctx, transport, downloader)
+			if err != nil {
+				d.downloadErr = log.Errorf("failed to get wasm: %w", err)
+				return
+			}
+			defer r.Close()
+
+			b, err := io.ReadAll(r)
+			if err != nil {
+				d.downloadErr = log.Errorf("failed to read wasm: %w", err)
+				return
+			}
+
+			if len(b) == 0 {
+				d.downloadErr = log.Errorf("received empty wasm")
+				return
+			}
+
+			dialer, err := createDialer(ctx, b, transport)
+			if err != nil {
+				d.downloadErr = log.Errorf("failed to create dialer: %w", err)
+			}
+			d.dialer = dialer
+		}()
 	}
 
+	return d, nil
+}
+
+func createDialer(ctx context.Context, wasm []byte, transport string) (water.Dialer, error) {
 	cfg := &water.Config{
 		TransportModuleBin: wasm,
 		OverrideLogger:     slog.New(newLogHandler(log, transport)),
 	}
 
 	dialer, err := water.NewDialerWithContext(ctx, cfg)
 	if err != nil {
-		return nil, fmt.Errorf("failed to create dialer: %w", err)
+		return nil, log.Errorf("failed to create dialer: %w", err)
 	}
-
-	return &waterImpl{
-		raddr:          addr,
-		dialer:         dialer,
-		reportDialCore: reportDialCore,
-	}, nil
+	return dialer, nil
 }
 
 func (d *waterImpl) dialServer(op *ops.Op, ctx context.Context) (net.Conn, error) {
 	return d.reportDialCore(op, func() (net.Conn, error) {
+		d.wgDownload.Wait()
```

Comment:
Hey @WendelHime sorry I just saw this logic.  Won't this lead to the client still getting blocked until the WASM download finishes?  The client will try to use this dialer, but the call to `waterImpl.dialServer` will block on this line.  I think you need to respect the input context, like:
```
select {
case <-someChannelWaitingOnWaterDownload:
    // carry on with dial
case <-ctx.Done():
    return nil, log.Errorf("context completed while waiting for WASM download: %w", ctx.Error())
}
```

However, even that has issues as the WATER dialer will then appear to be non-functional.  In the current code, the bandit dialer will simply [ignore it](https://github.com/getlantern/flashlight/blob/7fbdf0b33ce9bf0a8d13a7c66a90ab1d06b0b01d/bandit/bandit.go#L208-L213).  Even if that code changes, errors returned by a proxy dialer are always going to lead to that dialer being de-prioritized.  We don't want the WATER dialer to always get ignored or de-prioritized.

Instead, we need the WATER dialer to be added to the overall set of dialers only when it's ready for use.  I realize that is a more involved change.  Sorry that this comment is coming after this work was already reviewed and merged.

Reply:
🤔  can we introduce a new dialer function flag that says the dialer is ready? I'm considering these steps:

1. Create a `IsReady() bool` function at [bandit dialer interface](https://github.com/getlantern/flashlight/blob/main/bandit/bandit.go#L275).
2. Implement `IsReady` to all chained implementations.
3. Update water impl dialer to update the status properly.
4. Update [bandit logic for checking](https://github.com/getlantern/flashlight/blob/main/bandit/bandit.go#L207), if the dialer is not ready continue search of dialers.

This solve the issue for bandit, but not for bypass

Reply:
I like that approach!  Looks like bypass will get access to that function as well since `chained.CreateDialersMap` returns a map of `bandit.Dialers`.

Reply:
> In the current code, the bandit dialer will simply [ignore it](https://github.com/getlantern/flashlight/blob/7fbdf0b33ce9bf0a8d13a7c66a90ab1d06b0b01d/bandit/bandit.go#L208-L213). Even if that code changes, errors returned by a proxy dialer are always going to lead to that dialer being de-prioritized. We don't want the WATER dialer to always get ignored or de-prioritized.

Sorry, this had not occurred to me. Good catch!

Reply:
PR ready for review at: https://github.com/getlantern/flashlight/pull/1436